test/metabase/query_processor/pivot_test.clj

```diff
@@ -15,54 +16,167 @@
+         clojure.lang.ExceptionInfo
+         #"Invalid pivot-cols: specified breakout at index 3, but we only have 3 breakouts"
+         (#'pivot/breakout-combinations 3 [] [0 1 2 3]))))
+  ;; TODO -- we should require these columns to be distinct as well (I think?)
+  ;; TODO -- require all numbers to be positive
+  ;; TODO -- can you specify something in both pivot-rows and pivot-cols?
+  )
+
+(defn- test-query []
+  (mt/dataset sample-dataset
+    (mt/$ids orders
+      {:database     (mt/id)
+       :type         :query
+       :query        {:source-table $$orders
+                      :aggregation  [[:count]]
+                      :breakout     [$product_id->products.category
+                                     $user_id->people.source
+                                     !year.created_at]
+                      :filter       [:and
+                                     [:= $user_id->people.source "Facebook" "Google"]
+                                     [:= $product_id->products.category "Doohickey" "Gizmo"]
+                                     [:time-interval $created_at -2 :year {}]]}
+       :pivot-rows [0 1 2]
+       :pivot-cols []})))
+
+(deftest allow-snake-case-test
+  (testing "make sure the stuff works with either normal lisp-case keys or snake case. It should only ever see lisp-case"
+    (is (= (mt/rows (pivot/run-pivot-query (test-query)))
+           (mt/rows (pivot/run-pivot-query (set/rename-keys (test-query)
+                                                            {:pivot-rows :pivot_rows, :pivot-cols :pivot_cols})))))))
+
+(deftest dont-return-too-many-rows-test
+  (testing "Make sure pivot queries don't return too many rows (#14329)"
+    (mt/dataset sample-dataset
+      (let [results (pivot/run-pivot-query (test-query))
+            rows    (mt/rows results)]
+        ;; TODO - fix me
+        #_(is (= ["Category"
+                  "Source"
+                  "Created At"
+                  "pivot-grouping"
+                  "Count"]
+                 (map :display_name (mt/cols results))))
+        (is (apply distinct? rows))
+        ;; OLD RESULTS are commented out below
+        (is (= [;; ["Doohickey" "Facebook" "2019-01-01T00:00:00Z" 0  263 ]
+                ;; ["Doohickey" "Facebook" "2020-01-01T00:00:00Z" 0  89  ]
+                ;; ["Doohickey" "Google"   "2019-01-01T00:00:00Z" 0  276 ]
+                ;; ["Doohickey" "Google"   "2020-01-01T00:00:00Z" 0  100 ]
+                ;; ["Gizmo"     "Facebook" "2019-01-01T00:00:00Z" 0  361 ]
+                ;; ["Gizmo"     "Facebook" "2020-01-01T00:00:00Z" 0  113 ]
+                ;; ["Gizmo"     "Google"   "2019-01-01T00:00:00Z" 0  325 ]
+                ;; ["Gizmo"     "Google"   "2020-01-01T00:00:00Z" 0  101 ]
+                ;; [nil         nil        nil                    7  1628]
+                ["Doohickey" "Facebook" "2019-01-01T00:00:00Z" 0  263 ]
+                ["Doohickey" "Facebook" "2020-01-01T00:00:00Z" 0  89  ]
```

Comment:
I thought it looked weird that we were considering these duplicates? Unless I was misunderstanding.

Reply:
the rows that are commented out are the duplicates -- if you look at them they appear twice